### PR TITLE
Update release workflow

### PR DESCRIPTION
**Pull request by mikealfare**
# Description

This PR does multiple things that are all closely related:

### Migrate to `hatch`

Take the existing `dbt-athena-community` package and rename it `dbt-athena`.

- [x] move existing `dbt-athena-community` into `dbt-athena` namespace package
- [x] move code quality config into `.pre-commit-config.yaml`
- [x] create `pyproject.toml` for `dbt-athena`
- [x] remove `Makefile`, `MANIFEST.in`, `dev-requirements.txt`, `pytest.ini`, `setup.py`
- [x] update `ci`, `functional-tests`, and `functional-tests-workflow` workflows to point to the new targets

### Create `dbt-athena-community` shell package

Create a new `dbt-athena-community` package that is a simple wrapper around `dbt-athena` to preserve backwards compatibility.

- [x] create new `dbt-athena-community` namespace package
- [x] create `pyproject.toml` for `dbt-athena-community`
- [x] update `ci`, `functional-tests`, and `functional-tests-workflow` workflows to take the package as an input, defaulted to `dbt-athena`

### Update release

- [x] configure PyPI and GHA to use trusted publishers
- [x] create environments for publishing to test and prod
- [x] support publishing for both packages

### Standardize workflows

- [x] create reusable workflows for code-quality, unit-tests, integration-tests, and publish-pypi
- [x] replace ci with pull-request-checks, which calls to code-quality, unit-tests, and integration-tests
- [x] replace functional-tests-workflow with merge-checks, which calls to integration-tests
- [x] add label `ci:integration-tests` to optionally run integration tests during a pr (useful for not running on forks automatically, or skipping integration tests on small changes)
- [x] allow dispatch of all reusable workflows
- [x] use alls-green to minimize branch protection rules

### Update docs

- [x] update `CONTRIBUTING.md`
- [x] update`RELEASING.md`

# Checklist

- [x] You followed [contributing section](https://github.com/dbt-athena/dbt-athena#contributing)
- [x] You kept your Pull Request small and focused on a single feature or bug fix.
- [x] You added unit testing when necessary
- [x] You added functional testing when necessary

### CODE REVIEW

File: Makefile
Comment (nicor88) [inline Makefile:11]: be aware that this is used for the functional tests

File: Makefile
Comment (nicor88) [inline Makefile:1]: if you plan to remove totally the Makefile, it's not clear why, consider to update the CONTRIBUTING.md section.

IMHO Makefile are a really good starting point to create a set of commands for the contributors/developers - consider that when removing it, and create something equivalent ;)

File: Makefile
Comment (mikealfare) [reply Makefile:11]: Agreed, the functional test workflows will be updated to use the `hatch` scripts that do the same.

File: Makefile
Comment (mikealfare) [reply Makefile:1]: I'll update the docs, including `CONTRIBUTING.md` once the dust settles and I get all the workflows, releasing, etc. working. We're removing the Makefile because it's duplicative in functionality with `hatch` scripts.

We're moving to `hatch` across all adapters. We started with the `dbt-adapters` repo and the `dbt-postgres` repo since they were new. This felt like the right point here as well since we need to rewire some of the workflows for things like releasing. If you're unfamiliar with the tool, `hatch` feels a bit like a combination of `Makefile` and `tox`. We use a single default environment which is equivalent to `dev-requirements.txt` and we have scripts in there that are close to 1:1s with the `Makefile` recipes (which I can include in the docs). e.g.:
`make setup` -> `hatch run setup`
`make unit_test` -> `hatch run unit-tests`
`make functional_test` -> `hatch run integration-tests`
`make pre-commit-all` -> `hatch run code-quality`
Additionally, getting into the virtual environment is just `hatch shell`, similar to `python -m venv venv && source venv/bin/activate`.

File: Makefile
Comment (mikealfare) [reply Makefile:11]: I'm not sure if workflow runs are viewable, but if you can see [this](https://github.com/dbt-labs/dbt-athena/actions/runs/11276425220/job/31360167516), I got functional tests to work with `hatch`.

File: Makefile
Comment (mikealfare) [reply Makefile:1]: If you're interested, please take a look at the updated `CONTRIBUTING.md` [file](https://github.com/dbt-labs/dbt-athena/blob/actions/public-pypi-deploy/CONTRIBUTING.md) and let me know if this covers your concerns. I tried to capture each of the previous topics, but with the new tooling.

Comment (nicor88) [reply Makefile:1]: First of all, thanks for the context appreciate the exhaustive answer - I've been a big fan of Makefiles, and I was not aware of `hatch` - pretty cool. I had a look at the `CONTRIBUTING.md` and looks good to me. Thanks for addressing my concerns. Feel free to mark this as resolved, I know that your PR is in a draft state, I just wanted to raise my concerns:)

File: Makefile
Comment (nicor88) [reply Makefile:11]: All good here, you addressed my concerns on this https://github.com/dbt-labs/dbt-athena/pull/732#discussion_r1793927386

**Comment by mikealfare** [comment]
The skipped/failing `functional-tests` checks are the workflows from `main`. These don't work with the new setup since files moved and tests are invoked differently now. The same tests are run via `Pull request checks / integration-tests / integration-tests (pull_request)`.

I triggered the `Publish` workflow via the GH CLI with parameters for test PyPI and this branch:
```shell
gh workflow run publish.yml \
  --ref actions/public-pypi-deploy \
  -f deploy-to=test \
  -f branch=actions/public-pypi-deploy
```
Workflow run: https://github.com/dbt-labs/dbt-athena/actions/runs/11300329526
PyPI `dbt-athena`: https://test.pypi.org/project/dbt-athena/1.8.5rc1/
PyPI `dbt-athena-community`: https://test.pypi.org/project/dbt-athena-community/1.8.5rc1/

I can verify that installing `dbt-athena-community` gives me `dbt-athena`:

```shell
# create a virtual environment with only pip and setuptools installed
python -m venv venv
source venv/bin/activate
pip install --upgrade pip setuptools
pip list

Package    Version
---------- -------
pip        24.2
setuptools 75.1.0

# install dbt-athena-community from test PyPI to get the version in this PR
pip install dbt-athena-community==1.8.5rc1 \
  --index https://test.pypi.org/simple/ \
  --extra-index-url https://pypi.org/simple/

# we get both dbt-athena-community and dbt-athena
pip list | grep dbt-athena

dbt-athena                1.8.5rc1
dbt-athena-community      1.8.5rc1
```

Since `dbt-athena-community` used the namespace `athena` and not `athena-community`, everything should continue to work the same. The only difference in the packages is an extra, empty `athena-community` package inside `dbt/adapters` in the `dbt-athena-community` package.

File: .github/workflows/unit-tests.yml
Comment (colin-rogers-dbt) [inline .github/workflows/unit-tests.yml:42]: Might as well remove 3.8

File: .github/workflows/pull-request-checks.yml
Comment (colin-rogers-dbt) [inline .github/workflows/pull-request-checks.yml:30]: Do we need to gate running integration test behind a label?

Comment (mikealfare) [reply .github/workflows/pull-request-checks.yml:30]: That's what the repo originally had, and I don't hate it. What might also work is only running the workflow if the PR is in Ready for Review, but holding off if it's in Draft status. I think one of the reasons this existed is because some of the integration tests cannot run in parallel, which is why we also only run integration tests for one version of python. We should eventually figure out how to fix that. But in the meantime, if we have two PRs with changes happening in parallel, or scheduled tests running, etc. we could easily collide. This gives us a little more control. I can also remove it for now and address it separately.

Comment (nicor88) [reply .github/workflows/pull-request-checks.yml:30]: For context:

the reason why integration tests were under a label was simply because of cost - running integration tests concurrently was never an issue, because each test suite created a dedicated DB.

We didn't want to have the CI actions running when the PRs were in a draft or really early stage.
Generally after a first review, and then PR reached a "almost ready" to merge, me or another maintainer added the label to trigger the integration tests.

Comment (mikealfare) [reply .github/workflows/pull-request-checks.yml:30]: Thanks for the context @nicor88! Controlling cost makes sense. We don't have a control like that on other adapters, but maybe we should. I do like the more automatic nature of it depending on Draft vs Ready. @colin-rogers-dbt if you have concerns over the manual nature of adding the label, I can pull that for now and then we can address it separately if we want. Does that work?

> running integration tests concurrently was never an issue, because each test suite created a dedicated DB

Hmm, this is good to know. There's a single test that seems to always fail if I run on two versions of python in parallel. Perhaps CI is creating a DB per run, but not per matrix entry. It's probably worth looking at in a separate ticket.

File: dbt-athena-community/pyproject.toml
Comment (mikealfare) [inline dbt-athena-community/pyproject.toml:26]: Create a ticket to write a pre-commit hook that ensures these versions all match.

File: .github/workflows/merge-checks.yml
Comment (mikealfare) [inline .github/workflows/merge-checks.yml:1]: Re-purpose this for scheduled checks. We don't need checks on push since they run on PRs.

**Review by colin-rogers-dbt — DISMISSED** [review]

File: .github/workflows/merge-checks.yml
Comment (mikealfare) [reply .github/workflows/merge-checks.yml:1]: Add a ticket for this.

Comment (mikealfare) [reply .github/workflows/merge-checks.yml:1]: https://github.com/dbt-labs/dbt-athena/issues/742

File: .github/workflows/pull-request-checks.yml
Comment (mikealfare) [reply .github/workflows/pull-request-checks.yml:30]: I removed the label requirement.

File: dbt-athena-community/pyproject.toml
Comment (mikealfare) [reply dbt-athena-community/pyproject.toml:26]: https://github.com/dbt-labs/dbt-athena/issues/743

**Review by colin-rogers-dbt — APPROVED** [review]

**Pull request merged**
